In [1]:
import pandas as pd 

In [394]:
df= pd.read_csv('data_star_기대.csv')

In [395]:
df

Unnamed: 0         title directors  \
0             0            물괴       허종호   
1             1         싱글라이더       이주영   
2             2          허스토리       민규동   
3             3           풀잎들       홍상수   
4             4            영주       차성덕   
..          ...           ...       ...   
570         570          히말라야       이석훈   
571         571           허삼관       하정우   
572         572  해적: 바다로 간 산적       이석훈   
573         573           숫호구       백승기   
574         574          1987       장준환   

                                                actors  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   
..                                                 ...   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   

                                                 plots genre  runtime  \
0    “인왕산에 흉악한 짐승이 나타나 사람을 해쳤다 하옵니다.그것을 두고 듣지도 보지도 ...    액션    105.0   
1    그가 사라졌다그에게서 모든 것이 사라졌다증권회사의 지점장 강재훈(이병헌). 안정된 ...   드라마     97.0   
2     이겨야죠! 이겨야 할매들 분이 안풀리겠습니까? 1992~1998 6년의 기간, 2...   드라마    121.0   
3    커피집이 있을 것 같지 않은 골목 안으로 커피집이 있고 사람들이 커피집 안 여기저기...   드라마     66.0   
4    나쁘고 좋은 사람들을 만났다.교통사고로 부모를 잃고 졸지에 가장이 된 영주는 자신의...   드라마    101.0   
..                                                 ...   ...      ...   
570  “기다려... 우리가 꼭 데리러 갈게...”해발 8,750 미터 히말라야 에베레스트...   드라마    124.0   
571   그나저나 나한테 언제 시집 오실 생각이세요?  돈 없고, 대책 없고, 가진 것도 ...   드라마    123.0   
572  전국민을 시원하게 웃겨줄 액션 어드벤처!조선을 뒤흔든 최강도적들의 대격전! 조선의 ...  어드벤처    130.0   
573  얼굴, 몸매, 성적 매력까지 다 가진 슈퍼 섹시 아바타로 변신!내 인생 최고의 특급...   코미디     80.0   
574  “책상을 탁! 치니 억! 하고 죽었습니다”1987년 1월, 경찰 조사를 받던 스물두...   드라마    129.0   

                                              keywords        등급  국적  \
0                                조선시대,전염병,괴물,시대극/사극,사투  15세이상관람가  한국   
1                          기러기아빠,호주,오스트레일리아,반전,반전영화,동물  15세이상관람가  한국   
2    종군위안부,위안부,일본군위안부피해자,실화바탕,재판,관부재판,정신대,정신대할머니,법정...  12세이상관람가  한국   
3          삶과죽음,타인의 삶,언행,부산국제영화제,베를린영화제,삶의굴레,삶의순환,삶의의미  15세이상관람가  한국   
4                             부산국제영화제,부모님의죽음,장편데뷔작,복수극  12세이상관람가  한국   
..                                                 ...       ...  ..   
570  산,등산,히말라야,동료,에베레스트,실화바탕,실화바탕,실화바탕,실존인물,산악등반,동사...  12세이상관람가  한국   
571                                        소설원작,중국소설원작  12세이상관람가  한국   
572                                 고래,바다,사대부,조선,해적,국새  12세이상관람가  한국   
573                               백수,생체실험,정체불명,숫총각,아바타   청소년관람불가  한국   
574  518민주화운동,국가폭력,1980년대,1987년,대학생,검사,시위,신문기자,안기부,...  15세이상관람가  한국   

     screenCnt    audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0       1183.0   723414.0  6.104165e+09  2018-09-12  1.250000e+10    1274.0   
1        576.0   351276.0  2.685519e+09  2017-02-22  4.500000e+09    3838.0   
2        696.0   336822.0  2.521026e+09  2018-06-27  2.500000e+09     494.0   
3         54.0     7443.0  6.032025e+07  2018-10-25  1.000000e+08     162.0   
4        106.0    24823.0  1.927298e+08  2018-11-22  2.200000e+08     800.0   
..         ...        ...           ...         ...           ...       ...   
570     1009.0  7759473.0  6.017358e+10  2015-12-16  1.000000e+10   13693.0   
571      621.0   955175.0  7.405416e+09  2015-01-14  1.000000e+10   10731.0   
572      838.0  8646758.0  6.624063e+10  2014-08-06  1.500000e+10    4941.0   
573       14.0     1660.0  1.300650e+07  2014-08-07  5.000000e+06     247.0   
574     1299.0  7231638.0  5.816558e+10  2017-12-27  1.450000e+10   13877.0   

     star_buzz  
0         1629  
1          581  
2          943  
3          111  
4          335 

In [396]:
movie_buzz = pd.read_csv('movie_buzz_naver.csv')

In [397]:
df = pd.concat([df,movie_buzz], axis=1)

In [398]:
df

Unnamed: 0         title directors  \
0             0            물괴       허종호   
1             1         싱글라이더       이주영   
2             2          허스토리       민규동   
3             3           풀잎들       홍상수   
4             4            영주       차성덕   
..          ...           ...       ...   
570         570          히말라야       이석훈   
571         571           허삼관       하정우   
572         572  해적: 바다로 간 산적       이석훈   
573         573           숫호구       백승기   
574         574          1987       장준환   

                                                actors  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   
..                                                 ...   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   

                                                 plots genre  runtime  \
0    “인왕산에 흉악한 짐승이 나타나 사람을 해쳤다 하옵니다.그것을 두고 듣지도 보지도 ...    액션    105.0   
1    그가 사라졌다그에게서 모든 것이 사라졌다증권회사의 지점장 강재훈(이병헌). 안정된 ...   드라마     97.0   
2     이겨야죠! 이겨야 할매들 분이 안풀리겠습니까? 1992~1998 6년의 기간, 2...   드라마    121.0   
3    커피집이 있을 것 같지 않은 골목 안으로 커피집이 있고 사람들이 커피집 안 여기저기...   드라마     66.0   
4    나쁘고 좋은 사람들을 만났다.교통사고로 부모를 잃고 졸지에 가장이 된 영주는 자신의...   드라마    101.0   
..                                                 ...   ...      ...   
570  “기다려... 우리가 꼭 데리러 갈게...”해발 8,750 미터 히말라야 에베레스트...   드라마    124.0   
571   그나저나 나한테 언제 시집 오실 생각이세요?  돈 없고, 대책 없고, 가진 것도 ...   드라마    123.0   
572  전국민을 시원하게 웃겨줄 액션 어드벤처!조선을 뒤흔든 최강도적들의 대격전! 조선의 ...  어드벤처    130.0   
573  얼굴, 몸매, 성적 매력까지 다 가진 슈퍼 섹시 아바타로 변신!내 인생 최고의 특급...   코미디     80.0   
574  “책상을 탁! 치니 억! 하고 죽었습니다”1987년 1월, 경찰 조사를 받던 스물두...   드라마    129.0   

                                              keywords        등급  국적  \
0                                조선시대,전염병,괴물,시대극/사극,사투  15세이상관람가  한국   
1                          기러기아빠,호주,오스트레일리아,반전,반전영화,동물  15세이상관람가  한국   
2    종군위안부,위안부,일본군위안부피해자,실화바탕,재판,관부재판,정신대,정신대할머니,법정...  12세이상관람가  한국   
3          삶과죽음,타인의 삶,언행,부산국제영화제,베를린영화제,삶의굴레,삶의순환,삶의의미  15세이상관람가  한국   
4                             부산국제영화제,부모님의죽음,장편데뷔작,복수극  12세이상관람가  한국   
..                                                 ...       ...  ..   
570  산,등산,히말라야,동료,에베레스트,실화바탕,실화바탕,실화바탕,실존인물,산악등반,동사...  12세이상관람가  한국   
571                                        소설원작,중국소설원작  12세이상관람가  한국   
572                                 고래,바다,사대부,조선,해적,국새  12세이상관람가  한국   
573                               백수,생체실험,정체불명,숫총각,아바타   청소년관람불가  한국   
574  518민주화운동,국가폭력,1980년대,1987년,대학생,검사,시위,신문기자,안기부,...  15세이상관람가  한국   

     screenCnt    audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0       1183.0   723414.0  6.104165e+09  2018-09-12  1.250000e+10    1274.0   
1        576.0   351276.0  2.685519e+09  2017-02-22  4.500000e+09    3838.0   
2        696.0   336822.0  2.521026e+09  2018-06-27  2.500000e+09     494.0   
3         54.0     7443.0  6.032025e+07  2018-10-25  1.000000e+08     162.0   
4        106.0    24823.0  1.927298e+08  2018-11-22  2.200000e+08     800.0   
..         ...        ...           ...         ...           ...       ...   
570     1009.0  7759473.0  6.017358e+10  2015-12-16  1.000000e+10   13693.0   
571      621.0   955175.0  7.405416e+09  2015-01-14  1.000000e+10   10731.0   
572      838.0  8646758.0  6.624063e+10  2014-08-06  1.500000e+10    4941.0   
573       14.0     1660.0  1.300650e+07  2014-08-07  5.000000e+06     247.0   
574     1299.0  7231638.0  5.816558e+10  2017-12-27  1.450000e+10   13877.0   

     star_buzz  movie_buzz_naver  
0         1629                22  
1          581                

In [399]:
google_trend = pd.read_csv('google_trend.csv')

In [400]:
df = pd.merge(df,google_trend, how='left', on='title' )

In [401]:
df

Unnamed: 0         title directors  \
0             0            물괴       허종호   
1             1         싱글라이더       이주영   
2             2          허스토리       민규동   
3             3           풀잎들       홍상수   
4             4            영주       차성덕   
..          ...           ...       ...   
570         570          히말라야       이석훈   
571         571           허삼관       하정우   
572         572  해적: 바다로 간 산적       이석훈   
573         573           숫호구       백승기   
574         574          1987       장준환   

                                                actors  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   
..                                                 ...   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   

                                                 plots genre  runtime  \
0    “인왕산에 흉악한 짐승이 나타나 사람을 해쳤다 하옵니다.그것을 두고 듣지도 보지도 ...    액션    105.0   
1    그가 사라졌다그에게서 모든 것이 사라졌다증권회사의 지점장 강재훈(이병헌). 안정된 ...   드라마     97.0   
2     이겨야죠! 이겨야 할매들 분이 안풀리겠습니까? 1992~1998 6년의 기간, 2...   드라마    121.0   
3    커피집이 있을 것 같지 않은 골목 안으로 커피집이 있고 사람들이 커피집 안 여기저기...   드라마     66.0   
4    나쁘고 좋은 사람들을 만났다.교통사고로 부모를 잃고 졸지에 가장이 된 영주는 자신의...   드라마    101.0   
..                                                 ...   ...      ...   
570  “기다려... 우리가 꼭 데리러 갈게...”해발 8,750 미터 히말라야 에베레스트...   드라마    124.0   
571   그나저나 나한테 언제 시집 오실 생각이세요?  돈 없고, 대책 없고, 가진 것도 ...   드라마    123.0   
572  전국민을 시원하게 웃겨줄 액션 어드벤처!조선을 뒤흔든 최강도적들의 대격전! 조선의 ...  어드벤처    130.0   
573  얼굴, 몸매, 성적 매력까지 다 가진 슈퍼 섹시 아바타로 변신!내 인생 최고의 특급...   코미디     80.0   
574  “책상을 탁! 치니 억! 하고 죽었습니다”1987년 1월, 경찰 조사를 받던 스물두...   드라마    129.0   

                                              keywords        등급  국적  \
0                                조선시대,전염병,괴물,시대극/사극,사투  15세이상관람가  한국   
1                          기러기아빠,호주,오스트레일리아,반전,반전영화,동물  15세이상관람가  한국   
2    종군위안부,위안부,일본군위안부피해자,실화바탕,재판,관부재판,정신대,정신대할머니,법정...  12세이상관람가  한국   
3          삶과죽음,타인의 삶,언행,부산국제영화제,베를린영화제,삶의굴레,삶의순환,삶의의미  15세이상관람가  한국   
4                             부산국제영화제,부모님의죽음,장편데뷔작,복수극  12세이상관람가  한국   
..                                                 ...       ...  ..   
570  산,등산,히말라야,동료,에베레스트,실화바탕,실화바탕,실화바탕,실존인물,산악등반,동사...  12세이상관람가  한국   
571                                        소설원작,중국소설원작  12세이상관람가  한국   
572                                 고래,바다,사대부,조선,해적,국새  12세이상관람가  한국   
573                               백수,생체실험,정체불명,숫총각,아바타   청소년관람불가  한국   
574  518민주화운동,국가폭력,1980년대,1987년,대학생,검사,시위,신문기자,안기부,...  15세이상관람가  한국   

     screenCnt    audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0       1183.0   723414.0  6.104165e+09  2018-09-12  1.250000e+10    1274.0   
1        576.0   351276.0  2.685519e+09  2017-02-22  4.500000e+09    3838.0   
2        696.0   336822.0  2.521026e+09  2018-06-27  2.500000e+09     494.0   
3         54.0     7443.0  6.032025e+07  2018-10-25  1.000000e+08     162.0   
4        106.0    24823.0  1.927298e+08  2018-11-22  2.200000e+08     800.0   
..         ...        ...           ...         ...           ...       ...   
570     1009.0  7759473.0  6.017358e+10  2015-12-16  1.000000e+10   13693.0   
571      621.0   955175.0  7.405416e+09  2015-01-14  1.000000e+10   10731.0   
572      838.0  8646758.0  6.624063e+10  2014-08-06  1.500000e+10    4941.0   
573       14.0     1660.0  1.300650e+07  2014-08-07  5.000000e+06     247.0   
574     1299.0  7231638.0  5.816558e+10  2017-12-27  1.450000e+10   13877.0   

     star_buzz  movie_buzz_naver  google_trend  
0         1629                22         773.0  
1 

In [402]:
df['google_trend'] = df['google_trend'].fillna(0)

In [403]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 575 entries, 0 to 574
Data columns (total 19 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        575 non-null    int64  
 1   title             575 non-null    object 
 2   directors         575 non-null    object 
 3   actors            575 non-null    object 
 4   plots             575 non-null    object 
 5   genre             575 non-null    object 
 6   runtime           575 non-null    float64
 7   keywords          528 non-null    object 
 8   등급                575 non-null    object 
 9   국적                575 non-null    object 
 10  screenCnt         575 non-null    float64
 11  audiAcc           575 non-null    float64
 12  salesAcc          575 non-null    float64
 13  개봉일               575 non-null    object 
 14  제작비               575 non-null    float64
 15  네이버_기대지수          575 non-null    float64
 16  star_buzz         575 non-null    int64  
 1

In [217]:
#star = pd.read_csv('data_merge_star.csv')

In [218]:
#star.columns

In [219]:
#star = star.drop(['Unnamed: 0', 'directors', 'actors', 'plots', 'genre',
#       'runtime', 'keywords', '등급', '개봉일', '국적', 'screenCnt', 'salesAcc',
#       'audiAcc'], axis=1)

In [220]:
#star

In [221]:
#df = pd.merge(df,star, how='left', on='title' )

In [222]:
#df

In [223]:
#df_ect = pd.read_csv('data_merge_star.csv')

In [224]:
#df_1978 = df_ect[df_ect['title']=='1987']

In [225]:
#df_1978

In [226]:
#df = pd.concat([df,df_1978], axis=0)

In [227]:
#df.loc[df['title']=='1987','제작비'] = 14500000000

In [468]:
df =  pd.read_csv('Movie_최종.csv', encoding='cp949')

In [470]:
df = df.drop(['Column1','runtime','keywords','plots'], axis=1)

In [471]:
df

title directors  \
0              물괴       허종호   
1           싱글라이더       이주영   
2            허스토리       민규동   
3             풀잎들       홍상수   
4              영주       차성덕   
..            ...       ...   
570          히말라야       이석훈   
571           허삼관       하정우   
572  해적: 바다로 간 산적       이석훈   
573           숫호구       백승기   
574          1987       장준환   

                                                actors genre        등급  국적  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...    액션  15세이상관람가  한국   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   드라마  15세이상관람가  한국   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   드라마  12세이상관람가  한국   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   드라마  15세이상관람가  한국   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   드라마  12세이상관람가  한국   
..                                                 ...   ...       ...  ..   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   드라마  12세이상관람가  한국   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   드라마  12세이상관람가  한국   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...  어드벤처  12세이상관람가  한국   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   코미디   청소년관람불가  한국   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   드라마  15세이상관람가  한국   

     screenCnt  audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0         1183   723414  6.104165e+09  2018-09-12  1.250000e+10      1274   
1          576   351276  2.685519e+09  2017-02-22  4.500000e+09      3838   
2          696   336822  2.521026e+09  2018-06-27  2.500000e+09       494   
3           54     7443  6.032025e+07  2018-10-25  1.000000e+08       162   
4          106    24823  1.927298e+08  2018-11-22  2.200000e+08       800   
..         ...      ...           ...         ...           ...       ...   
570       1009  7759473  6.017358e+10  2015-12-16  1.000000e+10     13693   
571        621   955175  7.405416e+09  2015-01-14  1.000000e+10     10731   
572        838  8646758  6.624063e+10  2014-08-06  1.500000e+10      4941   
573         14     1660  1.300650e+07  2014-08-07  5.000000e+06       247   
574       1299  7231638  5.816558e+10  2017-12-27  1.450000e+10     13877   

     star_buzz  Movie_Buzz  
0         1629         795  
1          581         406  
2          943         646  
3          111           6  
4          335        1679  
..         ...         ...  
570       1987        1166  
571       1196        1010  
572       1413         379  
573        147         484  
574       1830        2654  

[575 rows x 14 columns]

# 타겟 만들기

In [472]:
df['Target'] = df['salesAcc'] - df['제작비']

In [473]:
df

title directors  \
0              물괴       허종호   
1           싱글라이더       이주영   
2            허스토리       민규동   
3             풀잎들       홍상수   
4              영주       차성덕   
..            ...       ...   
570          히말라야       이석훈   
571           허삼관       하정우   
572  해적: 바다로 간 산적       이석훈   
573           숫호구       백승기   
574          1987       장준환   

                                                actors genre        등급  국적  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...    액션  15세이상관람가  한국   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   드라마  15세이상관람가  한국   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   드라마  12세이상관람가  한국   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   드라마  15세이상관람가  한국   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   드라마  12세이상관람가  한국   
..                                                 ...   ...       ...  ..   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   드라마  12세이상관람가  한국   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   드라마  12세이상관람가  한국   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...  어드벤처  12세이상관람가  한국   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   코미디   청소년관람불가  한국   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   드라마  15세이상관람가  한국   

     screenCnt  audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0         1183   723414  6.104165e+09  2018-09-12  1.250000e+10      1274   
1          576   351276  2.685519e+09  2017-02-22  4.500000e+09      3838   
2          696   336822  2.521026e+09  2018-06-27  2.500000e+09       494   
3           54     7443  6.032025e+07  2018-10-25  1.000000e+08       162   
4          106    24823  1.927298e+08  2018-11-22  2.200000e+08       800   
..         ...      ...           ...         ...           ...       ...   
570       1009  7759473  6.017358e+10  2015-12-16  1.000000e+10     13693   
571        621   955175  7.405416e+09  2015-01-14  1.000000e+10     10731   
572        838  8646758  6.624063e+10  2014-08-06  1.500000e+10      4941   
573         14     1660  1.300650e+07  2014-08-07  5.000000e+06       247   
574       1299  7231638  5.816558e+10  2017-12-27  1.450000e+10     13877   

     star_buzz  Movie_Buzz        Target  
0         1629         795 -6.395835e+09  
1          581         406 -1.814481e+09  
2          943         646  2.102595e+07  
3          111           6 -3.967975e+07  
4          335        1679 -2.727025e+07  
..         ...         ...           ...  
570       1987        1166  5.017358e+10  
571       1196        1010 -2.594584e+09  
572       1413         379  5.124063e+10  
573        147         484  8.006500e+06  
574       1830        2654  4.366558e+10  

[575 rows x 15 columns]

In [474]:
df.loc[df['Target']<0,'Target'] = 0
df.loc[df['Target']>0,'Target'] = 1

In [475]:
df

title directors  \
0              물괴       허종호   
1           싱글라이더       이주영   
2            허스토리       민규동   
3             풀잎들       홍상수   
4              영주       차성덕   
..            ...       ...   
570          히말라야       이석훈   
571           허삼관       하정우   
572  해적: 바다로 간 산적       이석훈   
573           숫호구       백승기   
574          1987       장준환   

                                                actors genre        등급  국적  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...    액션  15세이상관람가  한국   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   드라마  15세이상관람가  한국   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   드라마  12세이상관람가  한국   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   드라마  15세이상관람가  한국   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   드라마  12세이상관람가  한국   
..                                                 ...   ...       ...  ..   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   드라마  12세이상관람가  한국   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   드라마  12세이상관람가  한국   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...  어드벤처  12세이상관람가  한국   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   코미디   청소년관람불가  한국   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   드라마  15세이상관람가  한국   

     screenCnt  audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0         1183   723414  6.104165e+09  2018-09-12  1.250000e+10      1274   
1          576   351276  2.685519e+09  2017-02-22  4.500000e+09      3838   
2          696   336822  2.521026e+09  2018-06-27  2.500000e+09       494   
3           54     7443  6.032025e+07  2018-10-25  1.000000e+08       162   
4          106    24823  1.927298e+08  2018-11-22  2.200000e+08       800   
..         ...      ...           ...         ...           ...       ...   
570       1009  7759473  6.017358e+10  2015-12-16  1.000000e+10     13693   
571        621   955175  7.405416e+09  2015-01-14  1.000000e+10     10731   
572        838  8646758  6.624063e+10  2014-08-06  1.500000e+10      4941   
573         14     1660  1.300650e+07  2014-08-07  5.000000e+06       247   
574       1299  7231638  5.816558e+10  2017-12-27  1.450000e+10     13877   

     star_buzz  Movie_Buzz  Target  
0         1629         795     0.0  
1          581         406     0.0  
2          943         646     1.0  
3          111           6     0.0  
4          335        1679     0.0  
..         ...         ...     ...  
570       1987        1166     1.0  
571       1196        1010     0.0  
572       1413         379     1.0  
573        147         484     1.0  
574       1830        2654     1.0  

[575 rows x 15 columns]

In [476]:
df.reset_index(inplace=True)

## 배우

In [477]:
df['act_n'] = df['actors'].str.split(',')

In [478]:
for i in range(len(df['act_n'])):
    df.loc[i,'act_n'] = len(df['act_n'][i])

## 배급사 합치기

In [479]:
df_pro = pd.read_excel('KOBIS_개봉일람.xlsx')

In [480]:
df_pro = df_pro.drop(['순번', '감독', '제작사', '수입사', '개봉일', '영화유형', '영화형태', '국적',
       '전국', '전국.1', '전국.2', '서울', '서울.1', '장르', '등급', '영화구분'], axis=1)

In [481]:
df_pro.rename(columns={'영화명':'title'}, inplace=True)

In [482]:
df_pro = df_pro.drop_duplicates()

In [483]:
df = pd.merge(df,df_pro, how='left', on='title' )

In [484]:
df = df.drop_duplicates('title', keep='first')

## 배급사 전처리

In [485]:
df['배급사']=df['배급사'].astype('str')

In [486]:
df[df['배급사']=='nan']

index title directors                                             actors  \
261    250    출국       노규엽  이범수, 연우진, 박혁권, 박주미, 이현정, 김보민, 로버트 미카, 전무송, 남문철...   
592    574  1987       장준환  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   

    genre        등급  국적  screenCnt  audiAcc      salesAcc         개봉일  \
261   드라마  15세이상관람가  한국        274    81722  6.395218e+08  2018-11-14   
592   드라마  15세이상관람가  한국       1299  7231638  5.816558e+10  2017-12-27   

              제작비  네이버_기대지수  star_buzz  Movie_Buzz  Target act_n  배급사  
261  6.500000e+09       150        427        1158     0.0    54  nan  
592  1.450000e+10     13877       1830        2654     1.0   295  nan

In [487]:
df.at[261, '배급사'] = '스파일이엔트'

df.at[592, '배급사'] = 'CJ'

In [488]:
df[df['배급사']=='nan']

Empty DataFrame
Columns: [index, title, directors, actors, genre, 등급, 국적, screenCnt, audiAcc, salesAcc, 개봉일, 제작비, 네이버_기대지수, star_buzz, Movie_Buzz, Target, act_n, 배급사]
Index: []

In [489]:
df['배급사'].value_counts()[:10] #160개

(주)씨제이이엔엠              78
(주)넥스트엔터테인먼트월드(NEW)    58
(주)쇼박스                 43
롯데쇼핑㈜롯데엔터테인먼트          39
CGV아트하우스               24
(주)리틀빅픽쳐스              23
롯데컬처웍스(주)롯데엔터테인먼트      21
메가박스중앙(주)플러스엠          21
㈜인디스토리                 16
(주)엣나인필름               11
Name: 배급사, dtype: int64

In [490]:
df['배급사'] = df['배급사'].str.replace('㈜','')

In [491]:
def get_dis(x) :
    if 'CJ' in x or 'CGV' in x or '씨제이' in x :
        return 'CJ'
    elif '쇼박스' in x :
        return '쇼박스'
    elif 'SK' in x or '에스케이' in x:
        return 'SK'
    elif '리틀빅픽' in x :
        return '리틀빅픽처스'
    elif '스폰지' in x :
        return '스폰지'
    elif '싸이더스' in x :
        return '싸이더스'
    elif '에이원' in x :
        return '에이원'
    elif '마인스' in x :
        return '마인스'
    elif '마운틴픽' in x :
        return '마운틴픽처스'
    elif '디씨드' in x :
        return '디씨드'
    elif '드림팩트' in x :
        return '드림팩트'
    elif '메가박스' in x :
        return '메가박스'
    elif '마운틴' in x :
        return '마운틴'
    elif '롯데' in x :
        return '롯데'
    elif 'KBS' in x :
        return 'KBS'
    elif '제이앤씨미디어그룹' in x :
        return '제이앤씨미디어그룹'
    elif '월트디즈니' in x :
        return '월트디즈니'
    elif '그린나래' in x :
        return '그린나래'
    elif 'KTG상상마당' in x or 'KT'in x :
        return 'KTG상상마당'
    elif 'TJOY' in x or 'T-JOY' in x:
        return 'TJOY'
    elif '디오시네마' in x :
        return '디오시네마'
    elif '스톰픽쳐스' in x :
        return '스톰픽쳐스'
    elif '디스테이션' in x :
        return '디스테이션'
    elif '더블앤조이' in x :
        return '디스테이션'
    elif 'CBS' in x :
        return 'CBS'
    elif 'CBS' in x :
        return 'CBS'
    elif '더쿱' in x :
        return '더쿱'
    elif '더콘텐츠온' in x :
        return '더콘텐츠온'
    elif '소니픽쳐스' in x :
        return '소니픽쳐스'
    elif '이수' in x :
        return '이수'
    elif '워너' in x :
        return '워너브라더스'
    elif '삼백' in x :
        return '삼백상회'
    elif '폭스' in x :
        return '폭스코리아'
    elif '엣나인' in x :
        return '엣나인필름'
    elif '티캐스트' in x :
        return '티캐스트'
    elif '콘텐츠판다' in x :
        return '콘텐츠판다'
    elif '판씨네마' in x :
        return '판씨네마'
    elif '키다리' in x :
        return '키다리이엔티'
    elif '스마일이' in x :
        return '스마일이엔티'
    elif '팝엔' in x :
        return '팝엔터테인먼트' 
    elif '오퍼스' in x :
        return '오퍼스픽쳐스' 
    elif '와이드' in x :
        return '와이드릴리즈' 
    elif '오드' in x :
        return '오드' 
    elif '조이앤' in x :
        return '조이앤시네마' 
    elif '박수' in x :
        return '박수엔터테인먼트' 
    elif '찬란' in x :
        return '찬란'
    elif '얼리' in x :
        return '얼리버드'
    elif '영화사빅' in x :
        return '영화사빅'
    elif '에이스' in x :
        return '에이스메이커'
    elif '퍼스트런' in x :
        return '퍼스트런'
    elif 'MM' in x or 'M&M' in x:
        return 'MM인터내셔널'
    elif '드림웨스트' in x :
        return '드림웨스트픽쳐스'
    elif '홀리가든' in x :
        return '홀리가든'
    elif 'R&R' in x :
        return 'RR프로덕션'
    elif '머스트' in x :
        return '머스트씨무비'
    elif '블룸즈' in x :
        return '블룸즈베리'
    elif '모쿠' in x :
        return '모쿠슈라'
    elif '니버설' in x :
        return '유니버설픽쳐스'
    elif '시네마달' in x :
        return '시네마달'
    else :
        return x

In [492]:
df['배급사'] = df['배급사'].apply(get_dis)

In [493]:
df['배급사'].nunique()

115

In [494]:
df['배급사'].value_counts()[:8]

CJ                     116
롯데                      61
(주)넥스트엔터테인먼트월드(NEW)     58
쇼박스                     45
리틀빅픽처스                  30
메가박스                    25
인디스토리                   16
엣나인필름                   13
Name: 배급사, dtype: int64

In [495]:
distributor_list = df['배급사'].value_counts()[:8]
def func(distributor):
    if distributor in distributor_list:
        return distributor
    else:
        return '기타'

In [496]:
df['배급사'] = df['배급사'].apply(lambda x : func(x))

In [497]:
tr_dist_rank =df.groupby('배급사').audiAcc.median().reset_index(name = 'dist_rank').sort_values(by = 'dist_rank',ascending=False)
tr_dist_rank

배급사  dist_rank
6                  쇼박스  2424330.0
0  (주)넥스트엔터테인먼트월드(NEW)  1535021.5
1                   CJ  1426615.0
3                   롯데  1110246.0
5                 메가박스   984265.0
4               리틀빅픽처스   112888.5
7                엣나인필름   103054.0
2                   기타     8680.0
8                인디스토리     2216.5

In [498]:
tr_dist_rank['dist_rank'] = [i + 1 for i in range(tr_dist_rank.shape[0])]
tr_dist_rank

배급사  dist_rank
6                  쇼박스          1
0  (주)넥스트엔터테인먼트월드(NEW)          2
1                   CJ          3
3                   롯데          4
5                 메가박스          5
4               리틀빅픽처스          6
7                엣나인필름          7
2                   기타          8
8                인디스토리          9

In [499]:
df = pd.merge(df, tr_dist_rank, how = 'left')

In [500]:
df

index         title directors  \
0        0            물괴       허종호   
1        1         싱글라이더       이주영   
2        2          허스토리       민규동   
3        3           풀잎들       홍상수   
4        4            영주       차성덕   
..     ...           ...       ...   
570    570          히말라야       이석훈   
571    571           허삼관       하정우   
572    572  해적: 바다로 간 산적       이석훈   
573    573           숫호구       백승기   
574    574          1987       장준환   

                                                actors genre        등급  국적  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...    액션  15세이상관람가  한국   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   드라마  15세이상관람가  한국   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   드라마  12세이상관람가  한국   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   드라마  15세이상관람가  한국   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   드라마  12세이상관람가  한국   
..                                                 ...   ...       ...  ..   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   드라마  12세이상관람가  한국   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   드라마  12세이상관람가  한국   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...  어드벤처  12세이상관람가  한국   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   코미디   청소년관람불가  한국   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   드라마  15세이상관람가  한국   

     screenCnt  audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0         1183   723414  6.104165e+09  2018-09-12  1.250000e+10      1274   
1          576   351276  2.685519e+09  2017-02-22  4.500000e+09      3838   
2          696   336822  2.521026e+09  2018-06-27  2.500000e+09       494   
3           54     7443  6.032025e+07  2018-10-25  1.000000e+08       162   
4          106    24823  1.927298e+08  2018-11-22  2.200000e+08       800   
..         ...      ...           ...         ...           ...       ...   
570       1009  7759473  6.017358e+10  2015-12-16  1.000000e+10     13693   
571        621   955175  7.405416e+09  2015-01-14  1.000000e+10     10731   
572        838  8646758  6.624063e+10  2014-08-06  1.500000e+10      4941   
573         14     1660  1.300650e+07  2014-08-07  5.000000e+06       247   
574       1299  7231638  5.816558e+10  2017-12-27  1.450000e+10     13877   

     star_buzz  Movie_Buzz  Target act_n                  배급사  dist_rank  
0         1629         795     0.0   121                   롯데          4  
1          581         406     0.0    50                   기타          8  
2          943         646     1.0    96  (주)넥스트엔터테인먼트월드(NEW)          2  
3          111           6     0.0    12                   기타          8  
4          335        1679     0.0    30                   CJ          3  
..         ...         ...     ...   ...                  ...        ...  
570       1987        1166     1.0    73                   CJ          3  
571       1196        1010     0.0    70  (주)넥스트엔터테인먼트월드(NEW)          2  
572       1413         379     1.0   148                   롯데          4  
573        147         484     1.0    32                엣나인필름          7  
574       1830        2654     1.0   295                   CJ          3  

[575 rows x 19 columns]

## 장르 전처리

In [501]:
df[df['genre']=='기타']

index  title directors                                            actors  \
185    185  인트로덕션       홍상수  신석호, 박미소, 김영호, 기주봉, 예지원, 조소연, 서영화, 김민희, 조윤희, 하성국   

    genre        등급  국적  screenCnt  audiAcc    salesAcc         개봉일  \
185    기타  12세이상관람가  한국         91     6185  55369270.0  2021-05-27   

            제작비  네이버_기대지수  star_buzz  Movie_Buzz  Target act_n 배급사  dist_rank  
185  50000000.0       284        142         348     1.0    10  기타          8

In [502]:
df.at[185, 'genre'] = '드라마'

In [503]:
df[df['genre']=='기타']

Empty DataFrame
Columns: [index, title, directors, actors, genre, 등급, 국적, screenCnt, audiAcc, salesAcc, 개봉일, 제작비, 네이버_기대지수, star_buzz, Movie_Buzz, Target, act_n, 배급사, dist_rank]
Index: []

In [504]:
df.groupby('genre').audiAcc.median().sort_values(ascending=False)

genre
어드벤처      8252669.0
사극        2022527.0
범죄        1736428.0
전쟁        1717566.0
액션        1638437.0
판타지       1535154.5
스릴러        616481.0
코미디        507257.0
미스터리       332993.0
멜로/로맨스     316780.0
공포(호러)     288016.0
SF         102170.0
드라마         95542.0
가족           8327.0
다큐멘터리        8138.5
뮤지컬          1027.0
Name: audiAcc, dtype: float64

In [505]:
df['genre_rank'] = df.genre.map({'뮤지컬' : 16,'다큐멘터리' : 15, '가족' : 14, '드라마' : 13, 'SF' : 12, '공포(호러)' : 11, '멜로/로맨스' : 10, '미스터리' : 9, '코미디':8,'스릴러':7,'판타지':6,'액션':5,'전쟁':4,'범죄':3,'사극':2,'어드벤처':1})

In [506]:
df

index         title directors  \
0        0            물괴       허종호   
1        1         싱글라이더       이주영   
2        2          허스토리       민규동   
3        3           풀잎들       홍상수   
4        4            영주       차성덕   
..     ...           ...       ...   
570    570          히말라야       이석훈   
571    571           허삼관       하정우   
572    572  해적: 바다로 간 산적       이석훈   
573    573           숫호구       백승기   
574    574          1987       장준환   

                                                actors genre        등급  국적  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...    액션  15세이상관람가  한국   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   드라마  15세이상관람가  한국   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   드라마  12세이상관람가  한국   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   드라마  15세이상관람가  한국   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   드라마  12세이상관람가  한국   
..                                                 ...   ...       ...  ..   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   드라마  12세이상관람가  한국   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   드라마  12세이상관람가  한국   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...  어드벤처  12세이상관람가  한국   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   코미디   청소년관람불가  한국   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   드라마  15세이상관람가  한국   

     screenCnt  audiAcc      salesAcc         개봉일           제작비  네이버_기대지수  \
0         1183   723414  6.104165e+09  2018-09-12  1.250000e+10      1274   
1          576   351276  2.685519e+09  2017-02-22  4.500000e+09      3838   
2          696   336822  2.521026e+09  2018-06-27  2.500000e+09       494   
3           54     7443  6.032025e+07  2018-10-25  1.000000e+08       162   
4          106    24823  1.927298e+08  2018-11-22  2.200000e+08       800   
..         ...      ...           ...         ...           ...       ...   
570       1009  7759473  6.017358e+10  2015-12-16  1.000000e+10     13693   
571        621   955175  7.405416e+09  2015-01-14  1.000000e+10     10731   
572        838  8646758  6.624063e+10  2014-08-06  1.500000e+10      4941   
573         14     1660  1.300650e+07  2014-08-07  5.000000e+06       247   
574       1299  7231638  5.816558e+10  2017-12-27  1.450000e+10     13877   

     star_buzz  Movie_Buzz  Target act_n                  배급사  dist_rank  \
0         1629         795     0.0   121                   롯데          4   
1          581         406     0.0    50                   기타          8   
2          943         646     1.0    96  (주)넥스트엔터테인먼트월드(NEW)          2   
3          111           6     0.0    12                   기타          8   
4          335        1679     0.0    30                   CJ          3   
..         ...         ...     ...   ...                  ...        ...   
570       1987        1166     1.0    73                   CJ          3   
571       1196        1010     0.0    70  (주)넥스트엔터테인먼트월드(NEW)          2   
572       1413         379     1.0   148                   롯데          4   
573        147         484     1.0    32                엣나인필름          7   
574       1830        2654     1.0   295                   CJ          3   

     genre_rank  
0             5  
1            13  
2            13  
3            13  
4            13  
..          ...  
570          13  
571          13  
572           1  
573           8  
574          13  

[575 rows x 20 columns]

## 등급 원핫(빼는게 좋은듯)

In [507]:
#df = pd.get_dummies(columns = ['등급'], data = df)

## 감독 전처리

In [508]:
df['directors'].nunique()

427

In [509]:
df['directors'].value_counts()[:70]

이준익    6
고봉수    4
류승완    4
이병헌    4
정지우    4
      ..
김지영    2
김봉한    2
김종관    2
창      2
장훈     2
Name: directors, Length: 70, dtype: int64

In [510]:
director_list = df['directors'].value_counts()[:100]
def cont_func(director):
    if director in director_list:
        return director
    else:
        return '기타'

In [511]:
df['directors'] = df['directors'].apply(lambda x : cont_func(x))

In [512]:
tr_direct_rank = df.groupby('directors').audiAcc.mean().reset_index(name = 'direct_rank').sort_values(by = 'direct_rank', ascending=False)
tr_direct_rank

directors   direct_rank
90       최동훈  1.284452e+07
26       김한민  1.254033e+07
41       봉준호  9.613966e+06
15       김용화  9.330964e+06
61       이석훈  8.203116e+06
..       ...           ...
75       이혁종  3.986000e+03
34       박석영  2.908000e+03
3        고봉수  2.345250e+03
38       백승기  2.281000e+03
84       정형석  9.650000e+02

[101 rows x 2 columns]

In [513]:
tr_direct_rank[:70]

directors   direct_rank
90       최동훈  1.284452e+07
26       김한민  1.254033e+07
41       봉준호  9.613966e+06
15       김용화  9.330964e+06
61       이석훈  8.203116e+06
..       ...           ...
72       이철하  8.399023e+05
6         기타  7.903726e+05
83       정지우  7.794115e+05
11    김선, 김곡  7.134430e+05
14       김용완  6.491440e+05

[70 rows x 2 columns]

In [514]:
tr_direct_rank['direct_rank'] = [i + 1 for i in range(tr_direct_rank.shape[0])]
tr_direct_rank[tr_direct_rank['directors'] == '기타']

directors  direct_rank
6        기타           67

In [515]:
df= pd.merge(df, tr_direct_rank, how = 'left')
df

index         title directors  \
0        0            물괴        기타   
1        1         싱글라이더        기타   
2        2          허스토리       민규동   
3        3           풀잎들       홍상수   
4        4            영주        기타   
..     ...           ...       ...   
570    570          히말라야       이석훈   
571    571           허삼관       하정우   
572    572  해적: 바다로 간 산적       이석훈   
573    573           숫호구       백승기   
574    574          1987        기타   

                                                actors genre        등급  국적  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...    액션  15세이상관람가  한국   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   드라마  15세이상관람가  한국   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   드라마  12세이상관람가  한국   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   드라마  15세이상관람가  한국   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   드라마  12세이상관람가  한국   
..                                                 ...   ...       ...  ..   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   드라마  12세이상관람가  한국   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   드라마  12세이상관람가  한국   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...  어드벤처  12세이상관람가  한국   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   코미디   청소년관람불가  한국   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   드라마  15세이상관람가  한국   

     screenCnt  audiAcc      salesAcc  ...           제작비  네이버_기대지수  star_buzz  \
0         1183   723414  6.104165e+09  ...  1.250000e+10      1274       1629   
1          576   351276  2.685519e+09  ...  4.500000e+09      3838        581   
2          696   336822  2.521026e+09  ...  2.500000e+09       494        943   
3           54     7443  6.032025e+07  ...  1.000000e+08       162        111   
4          106    24823  1.927298e+08  ...  2.200000e+08       800        335   
..         ...      ...           ...  ...           ...       ...        ...   
570       1009  7759473  6.017358e+10  ...  1.000000e+10     13693       1987   
571        621   955175  7.405416e+09  ...  1.000000e+10     10731       1196   
572        838  8646758  6.624063e+10  ...  1.500000e+10      4941       1413   
573         14     1660  1.300650e+07  ...  5.000000e+06       247        147   
574       1299  7231638  5.816558e+10  ...  1.450000e+10     13877       1830   

     Movie_Buzz  Target  act_n                  배급사 dist_rank  genre_rank  \
0           795     0.0    121                   롯데         4           5   
1           406     0.0     50                   기타         8          13   
2           646     1.0     96  (주)넥스트엔터테인먼트월드(NEW)         2          13   
3             6     0.0     12                   기타         8          13   
4          1679     0.0     30                   CJ         3          13   
..          ...     ...    ...                  ...       ...         ...   
570        1166     1.0     73                   CJ         3          13   
571        1010     0.0     70  (주)넥스트엔터테인먼트월드(NEW)         2          13   
572         379     1.0    148                   롯데         4           1   
573         484     1.0     32                엣나인필름         7           8   
574        2654     1.0    295                   CJ         3          13   

     direct_rank  
0             67  
1             67  
2             41  
3             94  
4             67  
..           ...  
570            5  
571           72  
572            5  
573          100  
574           67  

[575 rows x 21 columns]

In [516]:
df

index         title directors  \
0        0            물괴        기타   
1        1         싱글라이더        기타   
2        2          허스토리       민규동   
3        3           풀잎들       홍상수   
4        4            영주        기타   
..     ...           ...       ...   
570    570          히말라야       이석훈   
571    571           허삼관       하정우   
572    572  해적: 바다로 간 산적       이석훈   
573    573           숫호구       백승기   
574    574          1987        기타   

                                                actors genre        등급  국적  \
0    김명민, 김인권, 혜리, 최우식, 이경영, 박희순, 박성웅, 이도경, 김중희, 이규...    액션  15세이상관람가  한국   
1    이병헌, 공효진, 안소희, 잭 캠벨, 양유진, 애니카 화이틀리, 케이 에크룬드, 백...   드라마  15세이상관람가  한국   
2    김희애, 김해숙, 예수정, 문숙, 이용녀, 김선영, 김준한, 이유영, 이지하, 이용...   드라마  12세이상관람가  한국   
3    김민희, 정진영, 기주봉, 서영화, 김새벽, 안재홍, 공민정, 안선영, 신석호, 김...   드라마  15세이상관람가  한국   
4    김향기, 김호정, 유재명, 탕준상, 장혜진, 김용준, 강동윤, 박옥출, 현봉식, 전...   드라마  12세이상관람가  한국   
..                                                 ...   ...       ...  ..   
570  황정민, 정우, 조성하, 김인권, 라미란, 김원해, 이해영, 전배수, 정규수, 성병...   드라마  12세이상관람가  한국   
571  하정우, 하지원, 남다름, 노강민, 전현석, 전혜진, 장광, 민무제, 주진모, 성동...   드라마  12세이상관람가  한국   
572  손예진, 유해진, 김원해, 박철민, 조달환, 신정근, 설리, 이이경, 이경영, 김경...  어드벤처  12세이상관람가  한국   
573  백승기, 손이용, 박지나, 조한철, 김선호, 성충경, 신예주, 백대현, 이순자, 손...   코미디   청소년관람불가  한국   
574  김윤석, 하정우, 유해진, 김태리, 박희순, 이희준, 설경구, 여진구, 강동원, 유...   드라마  15세이상관람가  한국   

     screenCnt  audiAcc      salesAcc  ...           제작비  네이버_기대지수  star_buzz  \
0         1183   723414  6.104165e+09  ...  1.250000e+10      1274       1629   
1          576   351276  2.685519e+09  ...  4.500000e+09      3838        581   
2          696   336822  2.521026e+09  ...  2.500000e+09       494        943   
3           54     7443  6.032025e+07  ...  1.000000e+08       162        111   
4          106    24823  1.927298e+08  ...  2.200000e+08       800        335   
..         ...      ...           ...  ...           ...       ...        ...   
570       1009  7759473  6.017358e+10  ...  1.000000e+10     13693       1987   
571        621   955175  7.405416e+09  ...  1.000000e+10     10731       1196   
572        838  8646758  6.624063e+10  ...  1.500000e+10      4941       1413   
573         14     1660  1.300650e+07  ...  5.000000e+06       247        147   
574       1299  7231638  5.816558e+10  ...  1.450000e+10     13877       1830   

     Movie_Buzz  Target  act_n                  배급사 dist_rank  genre_rank  \
0           795     0.0    121                   롯데         4           5   
1           406     0.0     50                   기타         8          13   
2           646     1.0     96  (주)넥스트엔터테인먼트월드(NEW)         2          13   
3             6     0.0     12                   기타         8          13   
4          1679     0.0     30                   CJ         3          13   
..          ...     ...    ...                  ...       ...         ...   
570        1166     1.0     73                   CJ         3          13   
571        1010     0.0     70  (주)넥스트엔터테인먼트월드(NEW)         2          13   
572         379     1.0    148                   롯데         4           1   
573         484     1.0     32                엣나인필름         7           8   
574        2654     1.0    295                   CJ         3          13   

     direct_rank  
0             67  
1             67  
2             41  
3             94  
4             67  
..           ...  
570            5  
571           72  
572            5  
573          100  
574           67  

[575 rows x 21 columns]

In [517]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 575 entries, 0 to 574
Data columns (total 21 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   index        575 non-null    int64  
 1   title        575 non-null    object 
 2   directors    575 non-null    object 
 3   actors       575 non-null    object 
 4   genre        575 non-null    object 
 5   등급           575 non-null    object 
 6   국적           575 non-null    object 
 7   screenCnt    575 non-null    int64  
 8   audiAcc      575 non-null    int64  
 9   salesAcc     575 non-null    float64
 10  개봉일          542 non-null    object 
 11  제작비          575 non-null    float64
 12  네이버_기대지수     575 non-null    int64  
 13  star_buzz    575 non-null    int64  
 14  Movie_Buzz   575 non-null    int64  
 15  Target       575 non-null    float64
 16  act_n        575 non-null    object 
 17  배급사          575 non-null    object 
 18  dist_rank    575 non-null    int64  
 19  genre_ra

In [518]:
#df_fin = df.drop(['title','actors','배급사','genre','국적','directors','개봉일','audiAcc','salesAcc','등급','index','제작비'], axis=1)

In [519]:
df_fin = df.drop(['screenCnt','title','actors','배급사','genre','국적','directors','개봉일','audiAcc','salesAcc','index','제작비','등급'], axis=1)

In [520]:
df_fin['act_n'] = df_fin['act_n'].astype(int)

In [521]:
df_fin['Target']=df_fin['Target'].astype(int)

In [522]:
df_fin.corr()

네이버_기대지수  star_buzz  Movie_Buzz    Target     act_n  dist_rank  \
네이버_기대지수     1.000000   0.456861    0.128533  0.223013  0.356361  -0.262325   
star_buzz    0.456861   1.000000    0.248799  0.436474  0.501961  -0.576039   
Movie_Buzz   0.128533   0.248799    1.000000  0.163209  0.168155  -0.257606   
Target       0.223013   0.436474    0.163209  1.000000  0.182100  -0.399118   
act_n        0.356361   0.501961    0.168155  0.182100  1.000000  -0.334343   
dist_rank   -0.262325  -0.576039   -0.257606 -0.399118 -0.334343   1.000000   
genre_rank  -0.159123  -0.390730   -0.080575 -0.219618 -0.323091   0.354313   
direct_rank -0.239277  -0.599283   -0.091839 -0.313062 -0.372314   0.392187   

             genre_rank  direct_rank  
네이버_기대지수      -0.159123    -0.239277  
star_buzz     -0.390730    -0.599283  
Movie_Buzz    -0.080575    -0.091839  
Target        -0.219618    -0.313062  
act_n         -0.323091    -0.372314  
dist_rank      0.354313     0.392187  
genre_rank     1.000000     0.316747  
direct_rank    0.316747     1.000000

In [370]:
df_fin['Target'].value_counts()

1    302
0    273
Name: Target, dtype: int64

In [523]:
#df_fin.to_csv('전처리_naver_google합침.csv', index=False)

In [371]:
#df_fin['act_n'] = df_fin['act_n'].astype('int')
#df_fin['act_n'] = np.log1p(df_fin['act_n']) #로그변환
#df_fin['screenCnt'] = df_fin['screenCnt'].astype('int')
#df_fin['screenCnt'] = np.log1p(df_fin['screenCnt']) #로그변환
#df_fin['dist_rank'] = df_fin['dist_rank'].astype('int')
#df_fin['dist_rank'] = np.log1p(df_fin['dist_rank']) #로그변환
#df_fin['genre_rank'] = df_fin['genre_rank'].astype('int')
#df_fin['genre_rank'] = np.log1p(df_fin['genre_rank']) #로그변환
#df_fin['direct_rank'] = df_fin['direct_rank'].astype('int')
#df_fin['direct_rank'] = np.log1p(df_fin['direct_rank']) #로그변환
#df_fin['star_buzz'] = df_fin['star_buzz'].astype('int')
#df_fin['star_buzz'] = np.log1p(df_fin['star_buzz']) #로그변환

# 모델링

## train_test_split

In [592]:
df_fin = pd.read_csv('전처리_naver_movie_buzz_googletrend.csv')

In [593]:
y = df_fin['Target']
x = df_fin.drop('Target', axis=1)

In [594]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=100)

In [460]:
# 정규화
from sklearn.preprocessing import StandardScaler
 
sc = StandardScaler()
sc.fit(X_train)
 
X_train_std = sc.transform(X_train)
X_test_std = sc.transform(X_test)

## pycaret

In [77]:
from pycaret.classification import *
from sklearn.model_selection import train_test_split
train, test = train_test_split(df_fin, test_size=0.3, random_state=100)

In [375]:
setup_clf = setup(data = train, target= 'Target', fold_shuffle=True)

In [376]:
top10 = compare_models(sort='Accuracy', n_select=10, fold=5)

In [377]:
tuned_top10 = [tune_model(i, fold=5) for i in top10]

In [378]:
blender_top10 = blend_models(estimator_list=tuned_top10)

In [379]:
final_model = finalize_model(blender_top10)
prediction = predict_model(final_model, data=test)

In [380]:
prediction['Target'] = prediction['Target'].astype(int)

In [381]:
prediction['Label'] = prediction['Label'].astype(float)

In [382]:
prediction['Label'] = prediction['Label'].astype(int)

In [383]:
from pycaret.utils import check_metric
check_metric(prediction['Target'], prediction['Label'], metric = 'Accuracy') 
#0.68 naver movie buzz 미포함 # 0.7225 네이버무비버즈 포함
# 구글트렌드 포함 0.7399
# 구글트렌드+ 네이버 합침 0.7341

0.7341

스크린수 포함 최고 정확도 0.73(무비버즈 미포함)
스크린수 미포함 최고 정확도 0.72(무비버즈 미포함)

## logistic regression

In [87]:
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [387]:
model = LogisticRegression(penalty = 'l2')
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
accuracy_score(y_pred, y_test) 
# 구글트렌드 미포함 0.7225, 포함0.716
# 네이버 구글트렌드 합침 0.7109

0.7109826589595376

## gradient tree boosting

In [89]:
from sklearn.ensemble import GradientBoostingClassifier
from sklearn.metrics import roc_auc_score
from sklearn.model_selection import train_test_split

In [388]:
gbc = GradientBoostingClassifier(random_state=100, max_depth=3, learning_rate=0.01) 
gbc.fit(X_train, y_train)
score_train = gbc.score(X_train, y_train)
score_test = gbc.score(X_test, y_test)
gbc_roc_score = roc_auc_score(y_test, gbc.predict(X_test), average='macro')
print('ROC AUC: {0:.4f}'.format(gbc_roc_score))
print('정확도{:.4f}'.format(accuracy_score(y_pred, y_test)))
#구글트렌드 미포함 0.7225, 포함 0.7168
# 네이버 구글트렌드 합침 0.7110

ROC AUC: 0.7238
정확도0.7110


## linear discriminant

In [91]:
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.preprocessing import StandardScaler

In [389]:
lda = LinearDiscriminantAnalysis(n_components=1)
lda.fit(X_train, y_train)
#lda.transform(X_train)
y_pred = lda.predict(X_test)
accuracy_score(y_pred, y_test)
#구글트렌드 미포함 0.6936, 포함 0.6994
# 네이버 구글트렌드 합침 0.6878

0.6878612716763006

## random forest

In [93]:
from sklearn.ensemble import RandomForestClassifier

In [390]:
rfc = RandomForestClassifier(random_state=100)
rfc.fit(X_train, y_train)
y_pred = rfc.predict(X_test)
print('Model accuracy score with 10 decision-trees : {0:0.4f}'. format(accuracy_score(y_test, y_pred)))
# 구글트렌드 미포함 0.7283, 포함 0.7283
# 네이버 구글트렌드 합침 0.7110

Model accuracy score with 10 decision-trees : 0.7110


## catboost classifier

In [95]:
from catboost import CatBoostClassifier

In [391]:
clf = CatBoostClassifier(
    iterations=200,
    random_seed = 100,
    learning_rate=0.2, 
    loss_function='Logloss', #CrossEntropy,
    custom_metric = ['Logloss','AUC'],
    early_stopping_rounds = 20,
    use_best_model = True
)

In [97]:
params = {'iterations':200,
          'random_seed':63,
          'learning_rate':0.02,
          'loss_function':'Logloss', ## 사실 Default 값은 Logloss 이다. 만약 CatBoostRegressor 였으면, RMSE 이다.
          'custom_metric':['Logloss','AUC'],##, '
          'early_stopping_rounds':20,
          'use_best_model': True,
          'task_type':"GPU",
          'bagging_temperature':1,
          'verbose':False}

In [98]:
#cat_features = list(range(0, X_train.shape[1]))
#print(cat_features)

In [392]:
clf.fit(X_train, y_train, 
        #cat_features=cat_features, 
        eval_set=(X_test, y_test), 
        verbose=False
)

print('CatBoost model is fitted: ' + str(clf.is_fitted()))
print('CatBoost model parameters:')
print(clf.get_params())

CatBoost model is fitted: True
CatBoost model parameters:
{'iterations': 200, 'learning_rate': 0.2, 'loss_function': 'Logloss', 'random_seed': 100, 'use_best_model': True, 'custom_metric': ['Logloss', 'AUC'], 'early_stopping_rounds': 20}


In [393]:
y_pred = clf.predict(X_test)
print('정확도 : {0:0.4f}'. format(accuracy_score(y_test, y_pred)))
# 구글트렌드 미포함 0.7225. 포함 0.7399
# 네이버 구글트렌드 합침 0.7225

정확도 : 0.7225


## 보팅

In [101]:
from sklearn.ensemble import VotingClassifier

In [102]:
vo_clf = VotingClassifier(estimators=[('RF',rfc),('CAT',clf),('GTB',gbc),('LD',lda)], voting='soft')

In [439]:
vo_clf.fit(X_train, y_train)
pred = vo_clf.predict(X_test)
vo_roc_score = roc_auc_score(y_test, pred, average='macro')
print('ROC AUC: {0:.4f}'.format(vo_roc_score))
print("Voting 분류기 정확도", accuracy_score(y_test, pred))

RuntimeError: Cannot clone object <catboost.core.CatBoostClassifier object at 0x0000020D0DD6D940>, as the constructor either does not set or modifies parameter custom_metric

In [300]:
from xgboost import XGBClassifier
from sklearn.metrics import roc_auc_score
from sklearn.metrics import accuracy_score
# XGBClassifier 객체 생성
xgb_clf = XGBClassifier(n_estimators=100, random_state=156)

# 학습 : 성능 평가 지표를 auc로 설정하고 학습 수행. 
xgb_clf.fit(X_train, y_train, early_stopping_rounds=30,
            # 원래 검증 셋은 test 말고 다른 데이터 셋으로 해야하지만 여기서는 그냥 test 셋으로 진행함
            eval_metric="auc", eval_set=[(X_train, y_train), (X_test, y_test)]) #학습, 검증
#왼쪽이 학습하면서 정확도를 알려줌, 오른쪽은 학습한것에 대한 검증한것

[0]	validation_0-auc:0.90402	validation_1-auc:0.78313
[1]	validation_0-auc:0.94048	validation_1-auc:0.77946
[2]	validation_0-auc:0.96166	validation_1-auc:0.78422
[3]	validation_0-auc:0.96942	validation_1-auc:0.78129
[4]	validation_0-auc:0.97625	validation_1-auc:0.77946
[5]	validation_0-auc:0.98477	validation_1-auc:0.78245
[6]	validation_0-auc:0.98700	validation_1-auc:0.78748
[7]	validation_0-auc:0.98747	validation_1-auc:0.78918
[8]	validation_0-auc:0.98869	validation_1-auc:0.78429
[9]	validation_0-auc:0.99117	validation_1-auc:0.78510
[10]	validation_0-auc:0.99370	validation_1-auc:0.78544
[11]	validation_0-auc:0.99464	validation_1-auc:0.78571
[12]	validation_0-auc:0.99561	validation_1-auc:0.78680
[13]	validation_0-auc:0.99565	validation_1-auc:0.78816
[14]	validation_0-auc:0.99623	validation_1-auc:0.78803
[15]	validation_0-auc:0.99720	validation_1-auc:0.78340
[16]	validation_0-auc:0.99814	validation_1-auc:0.78653
[17]	validation_0-auc:0.99869	validation_1-auc:0.78653
[18]	validation_0-au

XGBClassifier(base_score=0.5, booster='gbtree', colsample_bylevel=1,
              colsample_bynode=1, colsample_bytree=1, enable_categorical=False,
              gamma=0, gpu_id=-1, importance_type=None,
              interaction_constraints='', learning_rate=0.300000012,
              max_delta_step=0, max_depth=6, min_child_weight=1, missing=nan,
              monotone_constraints='()', n_estimators=100, n_jobs=4,
              num_parallel_tree=1, objective='binary:logistic',
              predictor='auto', random_state=156, reg_alpha=0, reg_lambda=1,
              scale_pos_weight=1, subsample=1, tree_method='exact',
              use_label_encoder=True, validate_parameters=1, verbosity=None)

In [301]:
xgb_clf.predict_proba(X_test)[:,1]

array([0.18922459, 0.14622562, 0.9426745 , 0.76012886, 0.42007327,
       0.8997802 , 0.9372647 , 0.0591793 , 0.77322495, 0.70632654,
       0.7076092 , 0.04534187, 0.6917658 , 0.02642025, 0.41466248,
       0.9450615 , 0.9728098 , 0.9836031 , 0.8989627 , 0.68273854,
       0.17719446, 0.5088715 , 0.52080464, 0.94704163, 0.6237696 ,
       0.69877154, 0.02937572, 0.9027486 , 0.4749585 , 0.05086499,
       0.9314486 , 0.04100432, 0.5124604 , 0.44136027, 0.08965504,
       0.94875723, 0.0372237 , 0.915962  , 0.06306531, 0.87730306,
       0.08113704, 0.3869813 , 0.5452651 , 0.40517265, 0.6901556 ,
       0.04995109, 0.3440927 , 0.38334718, 0.7231264 , 0.12082391,
       0.9225146 , 0.78788865, 0.9863223 , 0.38581604, 0.11389484,
       0.80265445, 0.94522554, 0.8443156 , 0.02750267, 0.7542088 ,
       0.56743634, 0.13165274, 0.78968936, 0.20647205, 0.37934116,
       0.04101646, 0.95975333, 0.02106978, 0.76370454, 0.8779225 ,
       0.0234101 , 0.58950496, 0.8360975 , 0.5601416 , 0.46345

In [302]:
xgb_clf.predict(X_test) 

array([0, 0, 1, 1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 1, 0,
       1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 1, 0, 0, 0,
       1, 0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1])

In [303]:
xgb_roc_score = roc_auc_score(y_test, xgb_clf.predict(X_test))
print('ROC AUC: {0:.4f}'.format(xgb_roc_score))
# 구글트렌드 미포함 0.7265
# 포함 0.7207

ROC AUC: 0.7207


In [304]:
accuracy_score(y_test, xgb_clf.predict(X_test))
#구글트렌드 미포함 0.7167, 포함 0.7225
# 네이버 구글트렌드 합침 0.7225

0.7225433526011561

## DNN

In [108]:
#-*- coding: utf-8 -*-
from keras.models import Sequential
from keras.layers import Dense
from tensorflow.keras.layers import Dropout, Dense
from keras.callbacks import ModelCheckpoint,EarlyStopping
import tensorflow as tf
import os
from tensorflow.keras.layers import *
from tensorflow.keras.models import *
from tensorflow.keras.optimizers import *
from tensorflow.keras.callbacks import *

In [812]:
# 딥러닝 모델 선언
model = Sequential()
model.add(Dense(256, input_dim=8, activation='elu'))  # 입력층
model.add(BatchNormalization())
model.add(Dropout(0.2))
model.add(Dense(128, activation='elu'))      # 은닉층1 
model.add(BatchNormalization())
model.add(Dropout(0.2))
model.add(Dense(64, activation='elu'))      # 은닉층1
model.add(Dense(32, activation='elu'))
model.add(Dense(16, activation='elu'))
model.add(Dense(8, activation='elu'))
model.add(Dense(8, activation='elu'))
model.add(Dense(1))# 출력층
# 선형 회귀는 마지막에 참과 거짓을 구분할 필요가 없음. 출력층에 활성화 함수를 지정할 필요도 없음

model.compile(loss='binary_crossentropy',
              optimizer='rmsprop', 
              metrics=['acc'])

In [813]:
# 모델 저장 폴더 만들기
MODEL_DIR = './model/'
if not os.path.exists(MODEL_DIR):
    os.mkdir(MODEL_DIR)

modelpath = "./model/{epoch:02d}-{val_loss:.4f}.hdf5"

# 모델 업데이트 및 저장
checkpointer = ModelCheckpoint(filepath=modelpath, monitor='val_acc', verbose=0, save_best_only=True, mode='max')
# 콜백함수
reduce_lr = ReduceLROnPlateau(monitor='val_acc', factor=0.02, patience=100, mode='max')

# 학습 자동 중단 설정
early_stopping_callback = EarlyStopping(monitor='val_acc', patience=130, mode='max')

In [814]:
history = model.fit(x, y, validation_split=0.3, epochs=200, batch_size=32, verbose=1, callbacks=[early_stopping_callback,checkpointer,reduce_lr])
print("\ntrain acc: %.4f" % (model.evaluate(x, y)[1]))               

Epoch 1/500
26/26 [==============================] - 2s 14ms/step - loss: 5.2874 - acc: 0.5647 - val_loss: 3.6306 - val_acc: 0.6647 - lr: 0.0010
Epoch 2/500
26/26 [==============================] - 0s 5ms/step - loss: 3.7614 - acc: 0.6617 - val_loss: 2.4261 - val_acc: 0.7861 - lr: 0.0010
Epoch 3/500
26/26 [==============================] - 0s 4ms/step - loss: 5.5449 - acc: 0.5498 - val_loss: 4.1572 - val_acc: 0.5896 - lr: 0.0010
Epoch 4/500
26/26 [==============================] - 0s 4ms/step - loss: 4.5413 - acc: 0.5448 - val_loss: 2.8855 - val_acc: 0.7514 - lr: 0.0010
Epoch 5/500
26/26 [==============================] - 0s 4ms/step - loss: 2.7715 - acc: 0.6741 - val_loss: 2.1605 - val_acc: 0.7572 - lr: 0.0010
Epoch 6/500
26/26 [==============================] - 0s 4ms/step - loss: 2.6934 - acc: 0.6343 - val_loss: 2.6390 - val_acc: 0.5780 - lr: 0.0010
Epoch 7/500
26/26 [==============================] - 0s 4ms/step - loss: 2.3298 - acc: 0.6269 - val_loss: 1.3732 - val_acc: 0.6647 - lr

26/26 [==============================] - 0s 4ms/step - loss: 0.7627 - acc: 0.6617 - val_loss: 0.7958 - val_acc: 0.7514 - lr: 0.0010
Epoch 58/500
26/26 [==============================] - 0s 4ms/step - loss: 0.8406 - acc: 0.7015 - val_loss: 0.8339 - val_acc: 0.7861 - lr: 0.0010
Epoch 59/500
26/26 [==============================] - 0s 6ms/step - loss: 0.7799 - acc: 0.7239 - val_loss: 0.9781 - val_acc: 0.7630 - lr: 0.0010
Epoch 60/500
26/26 [==============================] - 0s 3ms/step - loss: 0.7558 - acc: 0.7239 - val_loss: 0.9337 - val_acc: 0.7514 - lr: 0.0010
Epoch 61/500
26/26 [==============================] - 0s 4ms/step - loss: 0.9621 - acc: 0.6940 - val_loss: 1.0088 - val_acc: 0.7225 - lr: 0.0010
Epoch 62/500
26/26 [==============================] - 0s 3ms/step - loss: 0.7536 - acc: 0.7015 - val_loss: 0.7460 - val_acc: 0.7688 - lr: 0.0010
Epoch 63/500
26/26 [==============================] - 0s 4ms/step - loss: 0.6583 - acc: 0.7214 - val_loss: 0.8062 - val_acc: 0.7514 - lr: 0.001

26/26 [==============================] - 0s 6ms/step - loss: 0.8687 - acc: 0.7388 - val_loss: 1.1643 - val_acc: 0.7688 - lr: 2.0000e-05
Epoch 114/500
26/26 [==============================] - 0s 5ms/step - loss: 0.8400 - acc: 0.7338 - val_loss: 1.2064 - val_acc: 0.7746 - lr: 2.0000e-05
Epoch 115/500
26/26 [==============================] - 0s 5ms/step - loss: 0.8755 - acc: 0.7562 - val_loss: 1.2063 - val_acc: 0.7746 - lr: 2.0000e-05
Epoch 116/500
26/26 [==============================] - 0s 5ms/step - loss: 0.8154 - acc: 0.7338 - val_loss: 1.2044 - val_acc: 0.7803 - lr: 2.0000e-05
Epoch 117/500
26/26 [==============================] - 0s 5ms/step - loss: 0.8579 - acc: 0.7463 - val_loss: 1.2040 - val_acc: 0.7746 - lr: 2.0000e-05
Epoch 118/500
26/26 [==============================] - 0s 6ms/step - loss: 0.7817 - acc: 0.7189 - val_loss: 1.2049 - val_acc: 0.7746 - lr: 2.0000e-05
Epoch 119/500
26/26 [==============================] - 0s 5ms/step - loss: 0.7617 - acc: 0.7786 - val_loss: 1.2031

#### 구글트렌드 미포함
0.7338 - elu, adam 256,normal, dropout, 128,normal, dropout, 64, 32, reduce-0.2,100, early-130, epoch-0.3,2000,128, auto
0.7488 - elu, adam 256,normal, dropout, 128,normal, dropout, 64, 32, reduce-0.2,100, early-130, epoch-0.3,2000,128, max
0.7587 - elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64, 32, reduce-0.2,100, early-130, epoch-0.3,2000,128, max
0.7836 0.7687
0.7512 - elu, rmsprop 32,normal, dropout, 128,normal, dropout, 64, 32, reduce-0.2,100, early-130, epoch-0.3,1000,128, max
0.7786,0.7811 - elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64, 32, 16, reduce-0.2,100, early-130, epoch-0.3,1000,128, max

#### 포함
0.8060- elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64, 32, 16, reduce-0.2,100, early-130, epoch-0.3,1000,128, max
0.7791- elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64, 32, 16, reduce-0.02,100, early-130, epoch-0.3,1000,128, max
0.7913- elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64, 32, 16, reduce-0.02,100, early-130, epoch-0.3,1000,32, max
0.8226,0.7826, 0.7983 elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64, 32, 16, 8, reduce-0.02,100, early-130, epoch-0.3,1000,32, max
0.7948,0.7757- elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64,normal, dropout, 32, 16, 8, reduce-0.02,100, early-130, epoch-0.3,1000,32, max
0.8104,0.8174- elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64,normal, dropout, 32, 16, 8,8, reduce-0.02,100, early-130, epoch-0.3,1000,32, max
#### 네이버 구글 합침
0.75- elu, rmsprop 256,normal, dropout, 128,normal, dropout, 64, 32, 16, reduce-0.2,100, early-130, epoch-0.3,1000,128, max

In [563]:
X = x.values

In [565]:
X = X[:,0:8]

In [550]:
from sklearn.model_selection import StratifiedKFold
skf = StratifiedKFold(n_splits = 10, random_state = 1000, shuffle = True)

In [ ]:
nn_acc = []
for train, test in skf.split(X,y):
    print(f'{i + 1} Fold Training.....')
    
    # 딥러닝 모델 선언
    model = Sequential()
    model.add(Dense(256, input_dim=8, activation='elu'))  # 입력층
    model.add(BatchNormalization())
    model.add(Dropout(0.2))
    model.add(Dense(128, activation='elu'))      # 은닉층1 
    model.add(BatchNormalization())
    model.add(Dropout(0.2))
    model.add(Dense(64, activation='elu'))      # 은닉층1
    model.add(Dense(32, activation='elu')) 
    model.add(Dense(16, activation='elu'))  
    model.add(Dense(8, activation='elu'))
    model.add(Dense(8, activation='elu'))
    model.add(Dense(1))# 출력층
    # 선형 회귀는 마지막에 참과 거짓을 구분할 필요가 없음. 출력층에 활성화 함수를 지정할 필요도 없음

    model.compile(loss='binary_crossentropy',optimizer='rmsprop', metrics=['acc'])

    # 모델 업데이트 및 저장
    mc = ModelCheckpoint(f'model_{i + 1}.h5', save_best_only = True, monitor = 'val_acc', mode = 'max', verbose = 0) #
    # 콜백함수
    reduce_lr = ReduceLROnPlateau(monitor='val_acc', factor=0.02, patience=100, mode='max')
    # 학습 자동 중단 설정
    early_stopping_callback = EarlyStopping(monitor='val_acc', patience=130, mode='max')

    #optimizer = Adam, RMSprop, Nadam, AdabelifOptimizer
    history = model.fit(X[train],y[train], validation_split=0.3, epochs=500, batch_size=32, verbose=0,callbacks=[early_stopping_callback,mc,reduce_lr]) 

    ### 최고 성능 기록 모델 Load
    best = load_model(f'model_{i + 1}.h5')
    ### validation predict
    val_pred = best.predict(X[test]).flatten()
    val_pred = np.where(val_pred>0.5, 1, 0)
    ### 확률값 중 최대값을 클래스로 매칭
    #val_cls = np.argmax(val_pred, axis = 1)
    ### Fold별 정확도 산출
    fold_nn_acc = accuracy_score(y[test], val_pred)
    nn_acc.append(fold_nn_acc)
    print(f'{i + 1} Fold ACC of NN = {fold_nn_acc}\n')

    ### Fold별 test 데이터에 대한 예측값 생성 및 앙상블
    #fold_pred = best.predict(X[test]) / skf.n_splits
    #nn_pred += fold_pred

In [810]:
nn_acc = []
for train, test in skf.split(X,y):
    print(f'{i + 1} Fold Training.....')
    
    # 딥러닝 모델 선언
    model = Sequential()
    model.add(Dense(256, input_dim=8, activation='elu'))  # 입력층
    model.add(BatchNormalization())
    model.add(Dropout(0.2))
    model.add(Dense(128, activation='elu'))      # 은닉층1 
    model.add(BatchNormalization())
    model.add(Dropout(0.2))
    model.add(Dense(64, activation='elu'))      # 은닉층1
    model.add(Dense(32, activation='elu')) 
    model.add(Dense(16, activation='elu'))  
    model.add(Dense(8, activation='elu'))
    model.add(Dense(8, activation='elu'))
    model.add(Dense(1))# 출력층
    # 선형 회귀는 마지막에 참과 거짓을 구분할 필요가 없음. 출력층에 활성화 함수를 지정할 필요도 없음

    model.compile(loss='binary_crossentropy',optimizer='rmsprop', metrics=['acc'])

    # 모델 업데이트 및 저장
    mc = ModelCheckpoint(f'model_{i + 1}.h5', save_best_only = True, monitor = 'val_acc', mode = 'max', verbose = 0) #
    # 콜백함수
    reduce_lr = ReduceLROnPlateau(monitor='val_acc', factor=0.02, patience=100, mode='max')
    # 학습 자동 중단 설정
    early_stopping_callback = EarlyStopping(monitor='val_acc', patience=130, mode='max')

    #optimizer = Adam, RMSprop, Nadam, Adabelif Optimizer
    history = model.fit(X[train],y[train], validation_split=0.3, epochs=500, batch_size=32, verbose=0,callbacks=[early_stopping_callback,mc,reduce_lr]) 

    ### 최고 성능 기록 모델 Load
    best = load_model(f'model_{i + 1}.h5')
    fold_nn_acc = best.evaluate(X[test], y[test])[1]         

    nn_acc.append(fold_nn_acc)
    print(f'{i + 1} Fold ACC of NN = {fold_nn_acc}\n')

    ### Fold별 test 데이터에 대한 예측값 생성 및 앙상블
    #fold_pred = best.predict(X[test]) / skf.n_splits
    #nn_pred += fold_pred

1 Fold Training.....
2/2 [==============================] - 0s 2ms/step - loss: 1.8257 - acc: 0.6552
1 Fold ACC of NN = 0.6551724076271057

1 Fold Training.....
2/2 [==============================] - 0s 3ms/step - loss: 1.2699 - acc: 0.7759
1 Fold ACC of NN = 0.7758620977401733

1 Fold Training.....
2/2 [==============================] - 0s 2ms/step - loss: 0.5831 - acc: 0.7414
1 Fold ACC of NN = 0.7413793206214905

1 Fold Training.....
2/2 [==============================] - 0s 4ms/step - loss: 2.0288 - acc: 0.6897
1 Fold ACC of NN = 0.6896551847457886

1 Fold Training.....
2/2 [==============================] - 0s 2ms/step - loss: 1.3635 - acc: 0.6724
1 Fold ACC of NN = 0.6724137663841248

1 Fold Training.....
2/2 [==============================] - 0s 3ms/step - loss: 1.7101 - acc: 0.7018
1 Fold ACC of NN = 0.7017543911933899

1 Fold Training.....
2/2 [==============================] - 0s 3ms/step - loss: 1.2891 - acc: 0.6667
1 Fold ACC of NN = 0.6666666865348816

1 Fold Training.....

In [811]:
np.mean(nn_acc)

0.71660618185997